### PR TITLE
in: fix #95 add an option to not download the s3 file

### LINKED ISSUE #95: [trigger] Add an option to not download the s3 file
Here is my usecase :

I build an application and put the file on s3
```
    - put: s3-release
      params:
        acl: private
        file: release/release.tar.gz
```

A second task is trigger by the get
```
    - get: s3-release
      trigger: true
      passed:
        - build-release
```

This task just need to know if the file have been change (we dont need to download it on concourse)

The issue is when we use get the file is downloaded

`967.73 MB / 967.73 MB [================================] 100.00 % 108.67 MB/s 8s`

Is there a way to have a parameter `skip_download` like docker-image-resource in order to just use s3 as a trigger ?

### PR DESCRIPTION
**Pull request by talset**
Adding skip_download param on the resource to not download file on
resource get.

Useful only trigger the pipeline without using the object.

### CODE REVIEW

**Comment by pivotal-issuemaster** [comment]
@talset Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/pivotal?repositoryId=concourse/s3-resource&pullRequestId=102)!

[Click here](https://cla.pivotal.io/sync/pivotal?repositoryId=concourse/s3-resource&pullRequestId=102) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

**Comment by pivotal-issuemaster** [comment]
@talset Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/pivotal?repositoryId=concourse/s3-resource&pullRequestId=102)!

File: in/command_test.go [hunk 1]
```diff
@@ -79,6 +79,17 @@ var _ = Describe("In Command", func() {
 			})
 		})
 
+		Context("when configured to skip download", func() {
+			BeforeEach(func() {
+				request.Source.SkipDownload = true
+			})
+
+			It("downloads the existing version of the file", func() {
```
Comment (vito) [inline in/command_test.go]: This says the opposite of what you want to demonstrate, and doesn't really assert anything. Could you reword this and add assertions that it doesn't download the file?

Something like `Expect(s3client.DownloadFileCallCount()).To(Equal(0))`.

File: in/command.go [hunk 2]
```diff
@@ -80,59 +80,60 @@ func (command *Command) Run(destinationDir string, request Request) (Response, e
...
-			if err != nil {
+			url = command.urlProvider.GetURL(request, remotePath)
+			if err = command.writeURLFile(destinationDir, url); err != nil {
 				return Response{}, err
 			}
 		}
 
-		url = command.urlProvider.GetURL(request, remotePath)
-		if err = command.writeURLFile(destinationDir, url); err != nil {
+		err = command.writeVersionFile(versionNumber, destinationDir)
```
Comment (vito) [inline in/command.go]: I don't think this should be skipped - downloading isn't required for writing the version file.

File: in/command.go [hunk 3]
```diff
@@ -80,59 +80,60 @@ func (command *Command) Run(destinationDir string, request Request) (Response, e
...
+				err = extractArchive(mime, destinationPath)
+				if err != nil {
+					return Response{}, err
+				}
 			}
-
-			err = extractArchive(mime, destinationPath)
-			if err != nil {
+			url = command.urlProvider.GetURL(request, remotePath)
+			if err = command.writeURLFile(destinationDir, url); err != nil {
```
Comment (vito) [inline in/command.go]: I don't think this should be skipped, either - they still may want the URL to the file even though they didn't download it.

File: in/command.go [hunk 4]
```diff
@@ -80,59 +80,60 @@ func (command *Command) Run(destinationDir string, request Request) (Response, e
 		isInitialVersion = request.Source.InitialVersion != "" && request.Version.VersionID == request.Source.InitialVersion
 	}
 
-	if isInitialVersion {
-		if request.Source.InitialContentText != "" || request.Source.InitialContentBinary == "" {
-			err = command.createInitialFile(destinationDir, path.Base(remotePath), []byte(request.Source.InitialContentText))
-			if err != nil {
-				return Response{}, err
+	if !request.Source.SkipDownload {
```
Comment (vito) [inline in/command.go]: I don't like the deep nesting this resulted in. Maybe just short-circuit in the `downloadFile` method instead?

Comment (talset) [reply in/command.go]: good point, we could eather just exclude the `downloadFile` and the `Unpack` part or modify the `downloadFile` method and update the `if request.Params.Unpack {` condition.

Update the `downloadFile` method could works but it feel strange to have a download method and put a parameter to not download.

WDYT ?

Comment (vito) [reply in/command.go]: True, we can just push the conditional down to wrap the download and unpack part instead of making the download method conditional.

**Comment by talset** [comment]
Thanks @vito, it have been updated with your changes request. Let me know it it is fine for you.

**Comment by vito** [comment]
thanks!

**Pull request merged**